runtime/middlewares.go

```diff
@@ -103,24 +110,59 @@ func (m *MiddlewareStack) Handle(
 	res *ServerHTTPResponse) context.Context {
 
 	shared := NewSharedState(m.middlewares)
+	start := time.Now()
 
 	for i := 0; i < len(m.middlewares); i++ {
 		ctx, ok := m.middlewares[i].HandleRequest(ctx, req, res, shared)
 		// If a middleware errors and writes to the response header
 		// then abort the rest of the stack and evaluate the response
 		// handlers for the middlewares seen so far.
 		if ok == false {
+			//record latency for middlewares requests in unsuccesful case as the middleware requests calls are terminated
+			m.recordLatency(middlewareRequestLatencyTag, start, req)
+
+			start = time.Now() // start the timer for middleware responses
 			for j := i; j >= 0; j-- {
 				m.middlewares[j].HandleResponse(ctx, res, shared)
 			}
+			//record latency for middlewares responses in unsuccesful case
+			m.recordLatency(middlewareResponseLatencyTag, start, req)
+
+			//for error metrics only emit when there is gateway error and not request error
+			if res.pendingStatusCode >= 500 {
```

Comment:
4xx is a success?

Reply:
I think it's the wrong place to verify statusCode. statusCode would be available only after we make call to downstream client - **ctx = m.handle(ctx, req, res)**

Reply:
> 4xx is a success?

Removed the success part as it can be taken from total_requests.Not taking 4xx as error is because the availability defined for gateway should not be dependent on the user's request and rather on the gateway internal errors, so that is why only 5xx.

Reply:
> I think it's the wrong place to verify statusCode. statusCode would be available only after we make call to downstream client - **ctx = m.handle(ctx, req, res)**

https://github.com/uber/zanzibar/blob/24bea8e47a8464853e5014da44ac30c2b17f091b/runtime/server_http_response.go#L271
We are setting the pendingStatuscode whenever we call the res.WriteJson from the middlewares,so this will always be populated in case of errors.